File: client/focus.lua
local focused = false

local function setFocus(state, cursor)
  focused = state
  SetNuiFocus(state, cursor == nil and state or cursor)
end

local function isFocused()
  return focused
end

exports('SetReactFocus', setFocus)
exports('IsReactFocused', isFocused)

RegisterNUICallback('react:setFocus', function(data, cb)
  setFocus(data.state == true, data.cursor)
  cb({ ok = true })
end)

RegisterNUICallback('react:close', function(data, cb)
  print(data.module)
  setFocus(false, false)
  cb({ ok = true, module = data and data.module })
end)

File: interface/src/lib/focus.ts
import { useEffect, useState } from 'react'
import { sendNui } from './nui'

export function useNuiFocus(initial = false) {
  const [focused, setFocused] = useState(initial)
  useEffect(() => {
    const onKey = (e: KeyboardEvent) => {
      if (!focused) return
      if (e.key === 'Escape') {
        sendNui('react:close')
        setFocused(false)
      }
    }
    window.addEventListener('keydown', onKey)
    return () => window.removeEventListener('keydown', onKey)
  }, [focused])
  useEffect(() => {
    sendNui('react:setFocus', { state: focused, cursor: focused })
  }, [focused])
  return { focused, setFocused }
}

